### PR TITLE
Postimplmentation

### PR DESCRIPTION
**Pull request by freta**
Done with all post api.

### CODE REVIEW

**Comment by leaderruvuna** [comment]
I've just seen your PR.I'm so tight and busy with zwaplace web application
that I'm working on right now and I must release the remaining tasks today.
So I'm gonna see whether I get time for review.I'll let you know.cheers

File: .babelrc [hunk 1]
```diff
@@ -5,5 +5,8 @@
             "node": "current"
          }
         }]
-    ]
+    ],
+  "plugins": [
+    "@babel/plugin-proposal-class-properties"
+  ]
 }
```
Comment (leaderruvuna) [inline .babelrc]: Make sure you have a new line at the end of every script you have. To enable GitHub utilities to operate on every file as required.

File: .eslintrc.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+        "Atomics": "readonly",
+        "SharedArrayBuffer": "readonly"
+    },
+    "parserOptions": {
+        "ecmaVersion": 2018,
+        "sourceType": "module"
+    },
+    "rules": {
+    }
+}
```
Comment (leaderruvuna) [inline .eslintrc.json]: Same thing for this file. Add a new line

File: .gitignore [hunk 3]
```diff
@@ -2,3 +2,5 @@ dist/
 node_modules/
 package-lock.json
 .vscode
+.env
+.vscode
```
Comment (leaderruvuna) [inline .gitignore]: Add new line.

File: schema.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    .required()
+    .message({
+        'string.base': 'Invalid type, username must be a string',
+        'string.empty': 'please enter your username',
+        'any.required': 'username is required',
+        'string.max': 'Username must be at most {#limit} characters long'
+    }),
+    email: joi.string().email().required()
+});
+export default schema;
```
Comment (leaderruvuna) [inline schema.js]: New line.

**Comment by leaderruvuna** [comment]
Copied that. I think one thing that I've realized within almost all your
scripts is that you forgot to add a new line at the end of each file you
have. I think you can do so and append updates. Thanks

File: src/app.js [hunk 5]
```diff
@@ -3,21 +3,22 @@ import express from "express";
 import * as bodyParser from "body-parser";
 import cors from "cors";
 
+
+//import {} from 'dotenv/config';
```
Comment (leaderruvuna) [inline src/app.js]: Remove all commented codes.

File: src/app.js [hunk 6]
```diff
@@ -3,21 +3,22 @@ import express from "express";
...
+require('dotenv').config({ path: ".env" });
 const app = express();
+const PORT = process.env.PORT || 8080;
+const db = require("./database/models");
+//require("./routes/auth.routes")(app);
```
Comment (leaderruvuna) [inline src/app.js]: Make sure there is no comment.

File: src/app.js [hunk 7]
```diff
@@ -3,21 +3,22 @@ import express from "express";
...
 const corsOptions = {
 origin : 'https://localhost:8081'
```
Comment (leaderruvuna) [inline src/app.js]: Put this URL in your .**env** file and access it from there.

File: src/app.js [hunk 8]
```diff
@@ -3,21 +3,22 @@ import express from "express";
...
 db.sequelize.sync();
-app.get("/", (req, res) => {
-  res.json({ message: "Welcome to level-up application." });
-});
-require("../src/routes/auth.routes")(app);
-const PORT = process.env.PORT || 8080;
+app.use(require("../src/routes/auth.routes"));
+app.use(require("../src/routes/user.routes"))
 app.listen(PORT, () => {
-  console.log("server is running");
```
Comment (leaderruvuna) [inline src/app.js:22]: Remove this console. Leave the other one showing the port.

File: src/controllers/auth.controller.js [hunk 9]
```diff
@@ -1,60 +1,42 @@
-import db from "../database/models";
-import config from "../database/config/auth.config";
-const Users = db.users;
```
Comment (leaderruvuna) [inline src/controllers/auth.controller.js:3]: Put the const Users after all imports.

File: src/controllers/auth.controller.js [hunk 10]
```diff
@@ -1,60 +1,42 @@
...
-import jwt from "jsonwebtoken";
-import bcrypt from "bcryptjs";
+import Response from '../helpers/response';
+import Password from '../helpers/password';
+import Token from '../helpers/token';
+import Userservices from '../database/services/userservice';
 
-exports.signup = (req, res) => {
-    console.log(req.body);
```
Comment (leaderruvuna) [inline src/controllers/auth.controller.js:8]: Before pushing your code make sure all consoles are cleared.

File: src/controllers/auth.controller.js [hunk 11]
```diff
@@ -1,60 +1,42 @@
...
-            username: user.username,
-            accessToken: jwt.sign({id:user.id}, config.secret, {expiresIn: 86400 })
-            });
-    }).catch(err => {
-        res.status(500).send({
-            message: err.message
-        });
-    });
-};
+class authControll{
```
Comment (leaderruvuna) [inline src/controllers/auth.controller.js]: Your class name should start with a capital.

File: src/controllers/auth.controller.js [hunk 12]
```diff
@@ -1,60 +1,42 @@
...
+static async signup(req, res){
```
Comment (leaderruvuna) [inline src/controllers/auth.controller.js]: Use camel case naming convention.

File: src/controllers/auth.controller.js [hunk 13]
```diff
@@ -1,60 +1,42 @@
...
+    const user = await Userservices.createUser(userData);
+    const token = Token.generateToken(user.user_id);
+    const {user_id} = user.dataValues;
+        return Response.handleSuccess(201,'Successful registered',{user_id,username,email,token},res);
+    }catch (err){
+        return Response.handleError(500,err.message,res);
+        }
+}
 
+static async signin(req,res){
```
Comment (leaderruvuna) [inline src/controllers/auth.controller.js]: Use camel case naming convention for all the methods that you're using.
Eg: **signUp()** && **signIn()** etc ...

File: src/controllers/auth.controller.js [hunk 14]
```diff
@@ -1,60 +1,42 @@
...
+   const token = Token.generateToken(user.user_id) 
+    return Response.handleSuccess(200,'Successful loggedIn',{username: user.username,token},res);
+}  catch (err) {
+    Response.handleError(500,err.message,res);
+   throw err;
+}
+}
+
+}
+export default  authControll;
```
Comment (leaderruvuna) [inline src/controllers/auth.controller.js]: Name authController like this : **AuthController**

File: src/controllers/auth.controller.js [hunk 15]
```diff
@@ -1,60 +1,42 @@
...
-            id: user.id,
-            username: user.username,
-            accessToken: token
-        });
-    }).catch(err=> {
-        res.status(500).send({
-            message: err.message
-        });
-    });
-};
```
Comment (leaderruvuna) [inline src/controllers/auth.controller.js:60]: Add new line.

File: src/app.js (hunk 5 above)
Comment (leaderruvuna) [inline src/app.js]: remove.

File: src/database/config/config.js [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+  },
+  production: {
+    username: process.env.DB_USERNAME,
+    password: process.env.DB_PASSWORD,
+    database: process.env.DB_DATABASE,
+    host: process.env.DB_HOST,
+    dialect: "mysql"
+  }
+}
+module.exports = config;
```
Comment (leaderruvuna) [inline src/database/config/config.js]: new line.

File: src/database/models/post.model.js [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+           },
+           updatedAt: {
+            type: Sequelize.DATE,
+            allowNull: false,
+            defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
+           }
+
+    });
+    return Posts;
+};
```
Comment (leaderruvuna) [inline src/database/models/post.model.js]: Add a new line.

File: src/database/models/user.model.js [hunk 18]
```diff
@@ -10,12 +15,21 @@ module.exports = (sequelize, Sequelize) => {
...
+           updatedAt: {
+            type: Sequelize.DATE,
+            allowNull: false,
+            defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
            }
     });
     return Users;
-};
\ No newline at end of file
+};
```
Comment (leaderruvuna) [inline src/database/models/user.model.js]: Add new line here also.

File: src/database/services/userservice.js [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+    if(!updatePost) return null;
+      return updatePost;
+}
+static createPost = async(post)=>{
+    const newPost = await Posts.create(post);
+    if(!newPost) return null;
+    return newPost;
+}
+}
+export default Userservices;
```
Comment (leaderruvuna) [inline src/database/services/userservice.js]: new line.

File: src/helpers/password.js [hunk 20]
```diff
@@ -0,0 +1,14 @@
...
+    return bcrypt.hashSync(pass, bcrypt.genSaltSync(8));
+    }
+static comparePass(pass,datapass){
+    return bcrypt.compareSync(
+            pass,
+            datapass
+        );
+    }
+}
+export default Password;
```
Comment (leaderruvuna) [inline src/helpers/password.js]: New line.

File: src/middlewares/index.js [hunk 21]
```diff
@@ -1,6 +1,6 @@
-import authjwt from "./authjwt";
-import verifySignUp from "./verifysignup";
+import authjwt from "./checkauth";
+import verifysignup from "./verifysignup";
 module.exports = {
     authjwt,
-    verifySignUp
+    verifysignup
 };
```
Comment (leaderruvuna) [inline src/middlewares/index.js]: Add new line

File: src/middlewares/validators.js [hunk 22]
```diff
@@ -0,0 +1,13 @@
...
+const signupValidator = (req, res, next)=>{
+try {
+    joi.assert(req.body, schema);
+} catch(err) {
+    return res.status(400).json({ err: err.details[0].message});
+}
+next();
+
+};
+export default signupValidator;
```
Comment (leaderruvuna) [inline src/middlewares/validators.js]: Add new line.

File: src/middlewares/verifysignup.js [hunk 23]
```diff
@@ -1,31 +1,12 @@
...
+const checkUsername = async (req, res, next) => {
+    const {username} = req.body;
+    const usernameExist = await Userservices.findUsername({username});
+    if (usernameExist){
+        return Response.handleError(400,"Username already in use!",res);
+    }
+    next();
 };
+
+export default checkUsername;
```
Comment (leaderruvuna) [inline src/middlewares/verifysignup.js]: new line.

File: src/middlewares/verifysignup.js [hunk 24]
```diff
@@ -1,31 +1,12 @@
-import db from "../database/models";
-const Users = db.users;
-module.exports.checkDuplicate = (req, res, next) => {
-    console.log(req.body);
```
Comment (leaderruvuna) [inline src/middlewares/verifysignup.js:4]: Remove console.log here.

File: src/routes/auth.routes.js [hunk 25]
```diff
@@ -1,12 +1,15 @@
-import { verifySignUp } from "../middlewares";
+import checkUsername from "../middlewares/verifysignup";
 import controller from "../controllers/auth.controller";
```
Comment (leaderruvuna) [inline src/routes/auth.routes.js]: Name the controller in a self-explanatory way like e.g: **UserController**  don't leave it raw like that it doesn't have any meaning.

File: src/routes/user.routes.js [hunk 26]
```diff
@@ -0,0 +1,19 @@
...
+        );
+        next();
+    });
+ 
+    router.get("/test/all", checkAuth, Authorization.allAcess);
+   router.get("/test/:id", checkAuth, Authorization.singlePost);
+   router.delete("/test/:id",checkAuth, Authorization.singlePostDelete);
+   router.put("/test/:id", checkAuth, Authorization.updatePost);
+   router.post("/test", checkAuth, Authorization.createPost);
+module.exports = router;
```
Comment (leaderruvuna) [inline src/routes/user.routes.js]: New Line. Make sure in all the files that you are working on must have a new line at the end of it to enable Github utilities of operating on files to work well.

**Review by leaderruvuna — COMMENTED** [review]
Make sure you read all the comments on each file based on what you should improve before merging your PR. After finishing all the changes you can use **git amend**  and then push for the new changes. So far your progress is fine. Keep it up, Thanks.

File: src/database/services/userservice.js [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+        });
+        if (!user) return null;
+    return user;
+}
+static createUser = async (user)=>{
+    const newUser = await Users.create(user);
+    if(!newUser) return null;
+    return newUser;
+}
+static allPosts = async (user_id)=>{
```
Comment (verdotte) [inline src/database/services/userservice.js]: create post service and put all the methods concerning post inside

File: src/helpers/password.js [hunk 28]
```diff
@@ -0,0 +1,14 @@
+import bcrypt from 'bcryptjs';
+
+class Password {
+static hashPassword(pass){
+    return bcrypt.hashSync(pass, bcrypt.genSaltSync(8));
+    }
+static comparePass(pass,datapass){
```
Comment (verdotte) [inline src/helpers/password.js]: indent your code well

File: src/routes/user.routes.js [hunk 29]
```diff
@@ -0,0 +1,19 @@
...
+const router = express.Router();
+
+ router.use((req, res, next) => {
+        res.header("Access-Control-Allow-Headers",
+        "x-access-token, Origin, Content-Type, Accept"
+        );
+        next();
+    });
+ 
+    router.get("/test/all", checkAuth, Authorization.allAcess);
```
Comment (verdotte) [inline src/routes/user.routes.js]: use `post`  as route url not `test` for more details follow the guideline on README file

File: src/controllers/user.controller.js [hunk 30]
```diff
@@ -0,0 +1,33 @@
+import Userservices from "../database/services/userservice";
+import Responses from "../helpers/response";
+
+class Authorization{
```
Comment (verdotte) [inline src/controllers/user.controller.js]: create post controller and put the post implementation inside

**Comment by freta** [comment]
I have made all the changes asked.

**Pull request merged**